src/shared/dns-domain.c

```diff
@@ -898,6 +898,75 @@ int dns_name_to_wire_format(const char *domain, uint8_t *buffer, size_t len, boo
         return out - buffer;
 }
 
+/* Decode a domain name according to RFC 1035 Section 3.1, without compression */
+int dns_name_from_wire_format(const uint8_t **data, size_t *len, char **ret) {
+        _cleanup_free_ char *domain = NULL;
+        const uint8_t *optval;
+        size_t optlen, n = 0;
+        int r;
+
+        assert(data);
+        assert(len);
+        assert(*data || *len == 0);
+        assert(ret);
+
+        optval = *data;
+        optlen = *len;
+
+        for (;;) {
+                const char *label;
+                uint8_t c;
+
+                /* Unterminated name */
+                if (optlen == 0)
+                        return -EBADMSG;
+
+                /* RFC 1035 § 3.1 total length of encoded name is limited to 255 octets */
+                if (*len - optlen > 255)
+                        return -EMSGSIZE;
+
+                c = *optval;
+                optval++;
+                optlen--;
+
+                if (c == 0)
+                        /* End label */
+                        break;
+                if (c > DNS_LABEL_MAX)
+                        return -EBADMSG;
+                if (c > optlen)
+                        return -EMSGSIZE;
+
+                /* Literal label */
+                label = (const char*) optval;
+                optval += c;
+                optlen -= c;
+
+                if (!GREEDY_REALLOC(domain, n + (n != 0) + DNS_LABEL_ESCAPED_MAX))
+                        return -ENOMEM;
+
+                if (n != 0)
+                        domain[n++] = '.';
+
+                r = dns_label_escape(label, c, domain + n, DNS_LABEL_ESCAPED_MAX);
```

Comment:
Looks like ASan isn't happy:
```sh
==167==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60f000133170 at pc 0x7f56a3b383ca bp 0x7ffc2f334ed0 sp 0x7ffc2f334ec8                       
READ of size 1 at 0x60f000133170 thread T0
    #0 0x7f56a3b383c9 in dns_label_escape /home/vagrant/systemd-asan/build/../src/shared/dns-domain.c:223:21
    #1 0x7f56a3b42cb5 in dns_name_from_wire_format /home/vagrant/systemd-asan/build/../src/shared/dns-domain.c:951:21
    #2 0x97d77a in ndisc_get_dns_name /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-router.c:834:13
    #3 0x97c35a in sd_ndisc_router_encrypted_dns_get_dnr /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-router.c:884:13
    #4 0x85e275 in ndisc_router_process_encrypted_dns /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1071:13
    #5 0x8485e0 in ndisc_router_process_options /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1152:29
    #6 0x83f368 in ndisc_router_handler /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1415:13
    #7 0x83d470 in ndisc_handler /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1445:21
    #8 0xa964d1 in ndisc_callback /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:42:9
    #9 0xa95832 in ndisc_handle_datagram /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:197:9
    #10 0xa924f9 in ndisc_recv /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:250:16
    #11 0x7f56a48a339c in source_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #12 0x7f56a48a13c4 in sd_event_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #13 0x7f56a48a6ebd in sd_event_run /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #14 0x7f56a48a7e49 in sd_event_loop /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #15 0x417339 in run /home/vagrant/systemd-asan/build/../src/network/networkd.c:112:13
    #16 0x41400d in main /home/vagrant/systemd-asan/build/../src/network/networkd.c:119:1
    #17 0x7f56a2e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)
    #18 0x7f56a2e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)
    #19 0x412a24 in _start (/usr/lib/systemd/systemd-networkd+0x412a24) (BuildId: dc92c90f76a3812a2510700974d528d4e5bab679)
0x60f000133170 is located 0 bytes after 176-byte region [0x60f0001330c0,0x60f000133170)
allocated by thread T0 here:
    #0 0x7f56a52eb1bc in __interceptor_calloc (/usr/lib64/clang/16/lib/libclang_rt.asan-x86_64.so+0xeb1bc) (BuildId: 5e54448b65ce364ffd26e5e37071cc7230e8931b)
    #1 0x96752a in ndisc_router_new /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-router.c:30:14
    #2 0xa910e4 in ndisc_recv /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:218:14
    #3 0x7f56a48a339c in source_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #4 0x7f56a48a13c4 in sd_event_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #5 0x7f56a48a6ebd in sd_event_run /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #6 0x7f56a48a7e49 in sd_event_loop /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #7 0x417339 in run /home/vagrant/systemd-asan/build/../src/network/networkd.c:112:13
    #8 0x41400d in main /home/vagrant/systemd-asan/build/../src/network/networkd.c:119:1
    #9 0x7f56a2e49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)
    #10 0x7f56a2e49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)
    #11 0x412a24 in _start (/usr/lib/systemd/systemd-networkd+0x412a24) (BuildId: dc92c90f76a3812a2510700974d528d4e5bab679)
SUMMARY: AddressSanitizer: heap-buffer-overflow /home/vagrant/systemd-asan/build/../src/shared/dns-domain.c:223:21 in dns_label_escape
Shadow bytes around the buggy address:
  0x60f000132e80: fd fd fd fd fa fa fa fa fa fa fa fa fd fd fd fd
  0x60f000132f00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x60f000132f80: fd fd fa fa fa fa fa fa fa fa fd fd fd fd fd fd
  0x60f000133000: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x60f000133080: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
=>0x60f000133100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00[fa]fa
  0x60f000133180: fa fa fa fa fa fa fd fd fd fd fd fd fd fd fd fd
  0x60f000133200: fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x60f000133280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x60f000133300: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                                                  [1040/1817]
  0x60f000133380: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==167==ABORTING
systemd-networkd.service: Main process exited, code=exited, status=1/FAILURE
```

Reply:
Do you happen to have the offending packet for this one? The logic looks okay here to me.

Reply:
I kind of sent more than a million nonsensical packets. I'm still looking for it

FWIW there appears to be another issue where functions from `src/shared/*` parsing domain names can return `EINVAL` (or anything other than `EBADMSG`) while `ndisc_router_process_options` expects `EBADMSG`. I'm not sure what to do though. (DNSSLs are affected too: https://github.com/systemd/systemd/pull/26079#issuecomment-1910206617). I'm looking for those packets too.

Reply:
The option was
```python
b'\x90\x06Bg\xae\x81\x1c\x07\x01\xd2<\xe9%\xf3!Kl\xb6L\xba\xc2\x1b2\xa4\x92\x13\r{v\x059\x88\xfa\xaaQ\xb6uME\x95\x85\x1e\x08\x00\x00\x00\x00\x00'
```
```sh
0000  90 06 42 67 AE 81 1C 07 01 D2 3C E9 25 F3 21 4B  ..Bg......<.%.!K
0010  6C B6 4C BA C2 1B 32 A4 92 13 0D 7B 76 05 39 88  l.L...2....{v.9.
0020  FA AA 51 B6 75 4D 45 95 85 1E 08 00 00 00 00 00  ..Q.uME.........
```
Looks like Wireshark can't parse it yet

Reply:
Thanks. Another oof, the length field was converted incorrectly.